File: apps/windows/SunoFlow/AccountManager.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using System.Threading.Tasks;

namespace SunoFlow;

/// <summary>Where this PC stands with a SunoFlow account.</summary>
internal enum AccountState
{
    NotConnected,
    Waiting,
    Connected,
    Failed,
}

/// <summary>
/// Links this PC to a SunoFlow account and holds the device key. The Windows
/// counterpart of <c>AccountManager.swift</c>, following the same device-
/// authorisation flow the website implements: ask for a code, send the user to
/// the browser to approve it, then poll until the key comes back. Nothing is
/// typed by the user and the app never handles a password — the browser does
/// the signing in.
///
/// The key is this PC's credential for the cleanup service, which is where a
/// dictation is checked against the user's trial or subscription. Without it
/// nothing authenticates and the app refuses to record, exactly as the macOS
/// app does.
///
/// It is stored DPAPI-encrypted under the current user account rather than as
/// plain text, so it is not readable by another user on the same machine and
/// does not survive being copied to a different one. That is the closest
/// Windows equivalent of the macOS Keychain available without dragging in a
/// credential-manager dependency.
/// </summary>
internal sealed class AccountManager
{
    public static readonly AccountManager Shared = new();

    private const string FunctionsBase = "https://asia-south1-sunoflow-app.cloudfunctions.net";
    private const string AccountUrl = "https://sunoflow-app.web.app/dashboard.html";

    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromSeconds(20) };
    private static readonly JsonSerializerOptions JsonOpts = new() { PropertyNameCaseInsensitive = true };

    private readonly object _gate = new();
    private CancellationTokenSource? _pairing;

    // The key is read on every dictation, so it is cached rather than decrypted
    // from disk each time.
    private string? _cachedKey;
    private bool _cacheLoaded;

    private AccountManager()
    {
        State = DeviceKey == null ? AccountState.NotConnected : AccountState.Connected;
    }

    /// <summary>Raised whenever <see cref="State"/> or the notice changes, so the
    /// settings window can redraw. Handlers must marshal to the UI thread.</summary>
    public event EventHandler? Changed;

    public AccountState State { get; private set; }

    /// <summary>The short code the user approves in the browser, while waiting.</summary>
    public string UserCode { get; private set; } = "";

    /// <summary>Why the last pairing attempt failed, when <see cref="State"/> is Failed.</summary>
    public string FailureMessage { get; private set; } = "";

    /// <summary>Set when the gateway last refused us for a subscription reason,
    /// so the settings window can say why instead of failing silently.</summary>
    public string? EntitlementNotice { get; private set; }

    public bool IsConnected => State == AccountState.Connected;

    /// <summary>This PC's device key, or null if it has never been connected.
    /// Read by the transcription client so the sidecar can authenticate to the
    /// cleanup gateway.</summary>
    public string? DeviceKey
    {
        get
        {
            lock (_gate)
            {
                if (!_cacheLoaded)
                {
                    _cachedKey = ReadKey();
                    _cacheLoaded = true;
                }
                return _cachedKey;
            }
        }
    }

    // --- Pairing ------------------------------------------------------------------

    /// <summary>Starts a pairing: fetches a code, opens the browser, then polls
    /// until the user approves it or it expires.</summary>
    public async Task ConnectAsync()
    {
        CancelPairing();
        var cts = new CancellationTokenSource();
        // Taken once. A CancellationToken stays usable after its source is
        // disposed, but registering a callback on one (which Task.Delay does)
        // does not — so this method, not CancelPairing, owns the disposal.
        var ct = cts.Token;
        lock (_gate) { _pairing = cts; }

        EntitlementNotice = null;
        SetState(AccountState.Waiting, code: "…");

        try
        {
            var pairing = await RequestPairingAsync(ct);
            ct.ThrowIfCancellationRequested();
            SetState(AccountState.Waiting, code: pairing.UserCode);

            // Hand off to the browser, where the user signs in.
            OpenInBrowser(pairing.VerificationUri);

            var key = await AwaitApprovalAsync(pairing, ct);
            ct.ThrowIfCancellationRequested();

            StoreKey(key);
            SetState(AccountState.Connected);
            AppLog.Log("account: device connected");
        }
        catch (OperationCanceledException)
        {
            // CancelPairing already set the state.
        }
        catch (PairingException ex)
        {
            SetState(AccountState.Failed, failure: ex.Message);
            AppLog.Log($"account: pairing failed — {ex.Message}");
        }
        catch (Exception ex)
        {
            SetState(AccountState.Failed, failure: "Couldn't connect. Check your internet and try again.");
            AppLog.Log($"account: pairing failed — {ex.Message}");
        }
        finally
        {
            lock (_gate)
            {
                if (ReferenceEquals(_pairing, cts)) _pairing = null;
            }
            cts.Dispose();
        }
    }

    public void CancelPairing()
    {
        CancellationTokenSource? cts;
        lock (_gate) { cts = _pairing; _pairing = null; }
        if (cts == null) return;
        // Cancel only. ConnectAsync disposes it in its finally, once it is
        // certain nothing is still waiting on the token.
        try { cts.Cancel(); } catch (ObjectDisposedException) { /* already finished */ }
        SetState(DeviceKey == null ? AccountState.NotConnected : AccountState.Connected);
    }

    /// <summary>Forgets the key on this PC. It does not revoke it on the server —
    /// only the account owner can do that, from the dashboard — so the wording in
    /// the UI has to be careful not to promise more than this does.</summary>
    public void SignOutThisPc()
    {
        CancelPairing();
        ClearKey();
        EntitlementNotice = null;
        SetState(AccountState.NotConnected);
        AppLog.Log("account: device key removed from this PC");
    }

    public void OpenAccountPage() => OpenInBrowser(AccountUrl);

    /// <summary>Called when the sidecar reports the gateway refused a dictation.</summary>
    public void NoteEntitlementProblem(string message)
    {
        EntitlementNotice = message;
        Changed?.Invoke(this, EventArgs.Empty);
    }

    public void ClearEntitlementNotice()
    {
        EntitlementNotice = null;
        Changed?.Invoke(this, EventArgs.Empty);
    }

    private void SetState(AccountState state, string code = "", string failure = "")
    {
        State = state;
        UserCode = code;
        FailureMessage = failure;
        Changed?.Invoke(this, EventArgs.Empty);
    }

    // --- Network ------------------------------------------------------------------

    private sealed class PairingException : Exception
    {
        public PairingException(string message) : base(message) { }
    }

    private sealed class Pairing
    {
        public string DeviceCode { get; init; } = "";
        public string UserCode { get; init; } = "";
        public string VerificationUri { get; init; } = "";
        public TimeSpan Interval { get; init; } = TimeSpan.FromSeconds(3);
        public DateTime ExpiresAt { get; init; }
    }

    private sealed class PairResponse
    {
        [JsonPropertyName("device_code")] public string DeviceCode { get; set; } = "";
        [JsonPropertyName("user_code")] public string UserCode { get; set; } = "";
        [JsonPropertyName("verification_uri")] public string VerificationUri { get; set; } = "";
        [JsonPropertyName("interval")] public double Interval { get; set; } = 3;
        [JsonPropertyName("expires_in")] public double ExpiresIn { get; set; } = 600;
    }

    private sealed class PollResponse
    {
        [JsonPropertyName("api_key")] public string ApiKey { get; set; } = "";
    }

    private static async Task<Pairing> RequestPairingAsync(CancellationToken ct)
    {
        var body = JsonSerializer.Serialize(new
        {
            name = Environment.MachineName,
            platform = "Windows " + Environment.OSVersion.Version,
            appVersion = typeof(AccountManager).Assembly.GetName().Version?.ToString() ?? "",
        });

        using var content = new StringContent(body, Encoding.UTF8, "application/json");
        using var resp = await Http.PostAsync($"{FunctionsBase}/pairDevice", content, ct);
        if (!resp.IsSuccessStatusCode)
            throw new PairingException("Couldn't start the connection. Check your internet and try again.");

        var json = await resp.Content.ReadAsStringAsync(ct);
        var parsed = JsonSerializer.Deserialize<PairResponse>(json, JsonOpts);
        if (parsed == null || string.IsNullOrEmpty(parsed.DeviceCode) || string.IsNullOrEmpty(parsed.UserCode))
            throw new PairingException("Couldn't start the connection. Check your internet and try again.");

        return new Pairing
        {
            DeviceCode = parsed.DeviceCode,
            UserCode = parsed.UserCode,
            VerificationUri = parsed.VerificationUri,
            Interval = TimeSpan.FromSeconds(Math.Max(1, parsed.Interval)),
            ExpiresAt = DateTime.UtcNow.AddSeconds(parsed.ExpiresIn),
        };
    }

    /// <summary>Polls until the user approves, the pairing expires, or we are cancelled.</summary>
    private static async Task<string> AwaitApprovalAsync(Pairing pairing, CancellationToken ct)
    {
        var body = JsonSerializer.Serialize(new { device_code = pairing.DeviceCode });

        while (DateTime.UtcNow < pairing.ExpiresAt)
        {
            ct.ThrowIfCancellationRequested();
            await Task.Delay(pairing.Interval, ct);
            ct.ThrowIfCancellationRequested();

            HttpResponseMessage resp;
            try
            {
                using var content = new StringContent(body, Encoding.UTF8, "application/json");
                resp = await Http.PostAsync($"{FunctionsBase}/pollDevice", content, ct);
            }
            catch (OperationCanceledException) when (ct.IsCancellationRequested)
            {
                throw;
            }
            catch
            {
                continue;   // transient network trouble: keep waiting
            }

            using (resp)
            {
                if ((int)resp.StatusCode == 428) continue;  // still waiting for approval
                if (!resp.IsSuccessStatusCode)
                    throw new PairingException("That connection expired. Start again to get a fresh code.");

                var json = await resp.Content.ReadAsStringAsync(ct);
                var parsed = JsonSerializer.Deserialize<PollResponse>(json, JsonOpts);
                if (parsed == null || string.IsNullOrEmpty(parsed.ApiKey))
                    throw new PairingException("The connection came back empty. Try again.");
                return parsed.ApiKey;
            }
        }
        throw new PairingException("The code expired before it was approved. Try again.");
    }

    private static void OpenInBrowser(string url)
    {
        if (string.IsNullOrEmpty(url)) return;
        try
        {
            Process.Start(new ProcessStartInfo(url) { UseShellExecute = true });
        }
        catch (Exception ex)
        {
            AppLog.Log($"account: couldn't open the browser — {ex.Message}");
        }
    }

    // --- Key storage (DPAPI) ------------------------------------------------------

    private static string KeyPath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "SunoFlow", "device.key");

    private static string? ReadKey()
    {
        try
        {
            if (!File.Exists(KeyPath)) return null;
            var plain = ProtectedData.Unprotect(
                File.ReadAllBytes(KeyPath), null, DataProtectionScope.CurrentUser);
            var value = Encoding.UTF8.GetString(plain);
            return string.IsNullOrWhiteSpace(value) ? null : value;
        }
        catch (Exception ex)
        {
            // A key encrypted by a different Windows user (or a corrupt file)
            // cannot be decrypted here. Treat it as absent so the app asks the
            // user to connect again rather than failing every dictation.
            AppLog.Log($"account: stored device key unreadable — {ex.Message}");
            return null;
        }
    }

    private void StoreKey(string key)
    {
        lock (_gate)
        {
            try
            {
                Directory.CreateDirectory(Path.GetDirectoryName(KeyPath)!);
                var blob = ProtectedData.Protect(
                    Encoding.UTF8.GetBytes(key), null, DataProtectionScope.CurrentUser);
                File.WriteAllBytes(KeyPath, blob);
                _cachedKey = key;
                _cacheLoaded = true;
            }
            catch (Exception ex)
            {
                AppLog.Log($"account: could not store the device key — {ex.Message}");
                throw new PairingException("Couldn't save the connection on this PC. Try again.");
            }
        }
    }

    private void ClearKey()
    {
        lock (_gate)
        {
            try { File.Delete(KeyPath); } catch { /* best-effort */ }
            _cachedKey = null;
            _cacheLoaded = true;
        }
    }
}

File: SunoFlowApp/Sources/SunoFlow/AccountManager.swift
import AppKit
import Foundation

/// Links this Mac to a SunoFlow account and holds the device key.
///
/// The pairing follows the device-authorisation flow the website implements:
/// ask for a code, send the user to the browser to approve it, then poll until
/// the key comes back. Nothing is typed by the user and the app never handles a
/// password — the browser does the signing in.
///
/// The key is the device's credential for the cleanup service. It lives in the
/// Keychain rather than UserDefaults so it is not readable from a plist by
/// anything that can read the user's home directory.
@MainActor
final class AccountManager: ObservableObject {
    static let shared = AccountManager()

    enum State: Equatable {
        case notConnected
        case waiting(code: String)
        case connected
        case failed(String)
    }

    @Published private(set) var state: State = .notConnected
    /// Set when the gateway last refused us for a subscription reason, so the
    /// dashboard can say why instead of failing silently.
    @Published private(set) var entitlementNotice: String?

    private static let functionsBase = URL(string: "https://asia-south1-sunoflow-app.cloudfunctions.net")!
    private static let accountURL = URL(string: "https://sunoflow-app.web.app/dashboard.html")!

    private var pollTask: Task<Void, Never>?

    private init() {
        state = Keychain.deviceKey() == nil ? .notConnected : .connected
    }

    /// The device key, if this Mac has been connected. Read by the transcription
    /// client so the sidecar can authenticate to the cleanup gateway.
    var deviceKey: String? { Keychain.deviceKey() }

    /// Derived from `state`, not from a fresh Keychain read. SwiftUI evaluates
    /// these on every render pass, and a Keychain hit per frame is how one
    /// password prompt becomes a hundred.
    var isConnected: Bool {
        if case .connected = state { return true }
        return false
    }

    // MARK: - Pairing

    func connect() {
        pollTask?.cancel()
        entitlementNotice = nil
        state = .waiting(code: "…")

        pollTask = Task { [weak self] in
            guard let self else { return }
            do {
                let pairing = try await self.requestPairing()
                guard !Task.isCancelled else { return }
                self.state = .waiting(code: pairing.userCode)

                // Hand off to the browser, where the user is already signed in.
                if let url = URL(string: pairing.verificationURI) {
                    NSWorkspace.shared.open(url)
                }

                let key = try await self.awaitApproval(pairing)
                guard !Task.isCancelled else { return }

                Keychain.setDeviceKey(key)
                self.state = .connected
                AppLog.log("account: device connected")
            } catch is CancellationError {
                // `cancel()` already set the state.
            } catch {
                guard !Task.isCancelled else { return }
                self.state = .failed((error as? PairingError)?.message ?? error.localizedDescription)
                AppLog.log("account: pairing failed — \(error.localizedDescription)")
            }
        }
    }

    func cancelPairing() {
        pollTask?.cancel()
        pollTask = nil
        state = isConnected ? .connected : .notConnected
    }

    /// Forgets the key on this Mac. It does not revoke it on the server — only
    /// the account owner can do that, from the dashboard — so the wording in the
    /// UI has to be careful not to promise more than this does.
    func signOutThisMac() {
        pollTask?.cancel()
        Keychain.clearDeviceKey()
        entitlementNotice = nil
        state = .notConnected
        AppLog.log("account: device key removed from this Mac")
    }

    func openAccountPage() {
        NSWorkspace.shared.open(Self.accountURL)
    }

    /// Called when the gateway refuses a cleanup for subscription reasons.
    func noteEntitlementProblem(_ message: String) {
        entitlementNotice = message
    }

    func clearEntitlementNotice() { entitlementNotice = nil }

    // MARK: - Network

    private struct Pairing {
        let deviceCode: String
        let userCode: String
        let verificationURI: String
        let interval: TimeInterval
        let expiresAt: Date
    }

    struct PairingError: Error {
        let message: String
    }

    private func requestPairing() async throws -> Pairing {
        var request = URLRequest(url: Self.functionsBase.appendingPathComponent("pairDevice"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 20
        request.httpBody = try JSONSerialization.data(withJSONObject: [
            "name": Host.current().localizedName ?? "Mac",
            "platform": "macOS " + ProcessInfo.processInfo.operatingSystemVersionString
                .replacingOccurrences(of: "Version ", with: ""),
            "appVersion": Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "",
        ])

        let (data, response) = try await URLSession.shared.data(for: request)
        guard (response as? HTTPURLResponse)?.statusCode == 200,
              let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let deviceCode = json["device_code"] as? String,
              let userCode = json["user_code"] as? String,
              let uri = json["verification_uri"] as? String
        else {
            throw PairingError(message: "Couldn't start the connection. Check your internet and try again.")
        }

        let interval = (json["interval"] as? Double) ?? 3
        let expiresIn = (json["expires_in"] as? Double) ?? 600
        return Pairing(
            deviceCode: deviceCode,
            userCode: userCode,
            verificationURI: uri,
            interval: max(1, interval),
            expiresAt: Date().addingTimeInterval(expiresIn)
        )
    }

    /// Polls until the user approves, the pairing expires, or we are cancelled.
    private func awaitApproval(_ pairing: Pairing) async throws -> String {
        var request = URLRequest(url: Self.functionsBase.appendingPathComponent("pollDevice"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = 20
        request.httpBody = try JSONSerialization.data(withJSONObject: ["device_code": pairing.deviceCode])

        while Date() < pairing.expiresAt {
            try Task.checkCancellation()
            try await Task.sleep(nanoseconds: UInt64(pairing.interval * 1_000_000_000))
            try Task.checkCancellation()

            guard let (data, response) = try? await URLSession.shared.data(for: request),
                  let http = response as? HTTPURLResponse
            else { continue }   // transient network trouble: keep waiting

            switch http.statusCode {
            case 200:
                guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let key = json["api_key"] as? String
                else { throw PairingError(message: "The connection came back empty. Try again.") }
                return key
            case 428:
                continue        // still waiting for the user to approve
            default:
                throw PairingError(message: "That connection expired. Start again to get a fresh code.")
            }
        }
        throw PairingError(message: "The code expired before it was approved. Try again.")
    }
}

// MARK: - Keychain

/// Minimal Keychain wrapper for the one secret this app stores.
///
/// Two things here matter more than they look.
///
/// **The value is cached in memory.** A Keychain read is not free and, more
/// importantly, an *untrusted* read shows the user a password prompt. The key is
/// read on every dictation and by SwiftUI properties that re-evaluate on each
/// render, so an uncached read turns one prompt into a storm of them.
///
/// **Writes carry an explicit ACL naming this application.** Without one, macOS
/// records the creating binary and prompts whenever a different one asks —
/// which is what happens the first time the app is moved, for example from a
/// build directory into /Applications.
enum Keychain {
    private static let service = "com.sunoapp.sunoflow"
    private static let account = "deviceKey"

    private static let lock = NSLock()
    private static var cached: String?
    private static var didRepair = false

    static func deviceKey() -> String? {
        lock.lock()
        defer { lock.unlock() }
        if let cached { return cached }
        let value = read()
        cached = value
        return value
    }

    private static func read() -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess,
              let data = item as? Data,
              let value = String(data: data, encoding: .utf8),
              !value.isEmpty
        else { return nil }
        return value
    }

    static func setDeviceKey(_ key: String) {
        lock.lock()
        defer { lock.unlock() }
        write(key)
        cached = key
    }

    private static func write(_ key: String) {
        deleteItem()
        var attributes: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecValueData as String: Data(key.utf8),
            // Available without an unlock prompt after first unlock, and never
            // synced to iCloud or included in a backup.
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        ]
        if let access = selfAccess() {
            attributes[kSecAttrAccess as String] = access
        }
        let status = SecItemAdd(attributes as CFDictionary, nil)
        if status != errSecSuccess {
            AppLog.log("account: keychain write failed (\(status))")
        }
    }

    /// An ACL that trusts this application, so reads do not prompt.
    ///
    /// These APIs are long-deprecated but remain the only way to set a keychain
    /// ACL for an unsandboxed, directly-distributed Mac app. The modern
    /// data-protection keychain needs an application-identifier entitlement,
    /// which requires provisioning this app deliberately does not use.
    private static func selfAccess() -> SecAccess? {
        var trusted: SecTrustedApplication?
        guard SecTrustedApplicationCreateFromPath(nil, &trusted) == errSecSuccess,
              let trusted
        else { return nil }
        var access: SecAccess?
        guard SecAccessCreate("SunoFlow device key" as CFString,
                              [trusted] as CFArray, &access) == errSecSuccess
        else { return nil }
        return access
    }

    /// Rewrites the stored key so its ACL names the binary that is running now.
    ///
    /// Called once at launch. An item written by a copy of the app at a
    /// different path — the usual case after installing into /Applications —
    /// prompts on every read until it is rewritten from the new location. The
    /// first read here may still prompt once; after that it is silent.
    static func repairAccessIfNeeded() {
        lock.lock()
        defer { lock.unlock() }
        guard !didRepair else { return }
        didRepair = true
        guard let existing = cached ?? read() else { return }
        write(existing)
        cached = existing
        AppLog.log("account: keychain entry rewritten for this copy of the app")
    }

    static func clearDeviceKey() {
        lock.lock()
        defer { lock.unlock() }
        deleteItem()
        cached = nil
    }

    private static func deleteItem() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(query as CFDictionary)
    }
}
